File: native/cpp/CentroidStrategy.cpp
#include "CentroidStrategy.h"
#include <iostream>

void CentroidUtility::initializeClusters(const Eigen::MatrixXd &X, Eigen::MatrixXd &centroids, int num_clusters, CentroidStrategy clusterStrategy) {
    centroids = Eigen::MatrixXd::Zero(X.rows(), num_clusters);
    //std::cout << X.rows() << "*" << num_clusters << std::endl;

    if (clusterStrategy == CentroidStrategy::MIRTILL) {
        for (int d = 0; d < num_clusters; ++d) {
            double val = (double) d / (num_clusters - 1);
            centroids.col(d).setConstant(val);
        }
    } else if (clusterStrategy == CentroidStrategy::IRIS_DIAGONAL) {
        
        /*
        for (int d = 0; d < X.rows(); ++d) {
            centroids(d, 0) = (1 - std::pow(-1, d)) / 2;
            centroids(d, 1) = 0.5;
            centroids(d, 2) = (1 + std::pow(-1, d)) / 2;
        }
        */
    } else if (clusterStrategy == CentroidStrategy::K_MEANS) {
        // nincs most arra idonk
    } else {
        throw std::runtime_error("Cluster strategy not implemented");
    }

    bool debug = false;

    if (debug) {
        std::cout << "Centroids matrix:" << std::endl;
        std::cout << centroids << std::endl;
    }
}

File: native/rust/algofuzz/src-nalgebra/centroid_strategy.rs
use nalgebra::{DMatrix, DVector};

pub enum CentroidStrategy {
    Mirtill,
    IrisDiagonal,
    KMeans,
}

impl CentroidStrategy {
    pub fn to_centroid_strategy(strategy: f64) -> CentroidStrategy {
        match strategy as i32 {
            0 => CentroidStrategy::Mirtill,
            1 => CentroidStrategy::IrisDiagonal,
            2 => CentroidStrategy::KMeans,
            _ => panic!("Invalid centroid strategy"),
        }
    }
}

pub struct CentroidUtility;

impl CentroidUtility {
    pub fn initialize_clusters(
        X: &DMatrix<f64>,
        centroids: &mut DMatrix<f64>,
        num_clusters: usize,
        cluster_strategy: &CentroidStrategy,
    ) {
        *centroids = DMatrix::zeros(X.nrows(), num_clusters);

        match cluster_strategy {
            CentroidStrategy::Mirtill => {
                for d in 0..num_clusters {
                    let val = d as f64 / (num_clusters - 1) as f64;
                    centroids.set_column(d, &DVector::from_element(X.nrows(), val));
                }
            }
            CentroidStrategy::IrisDiagonal => {
                for d in 0..X.nrows() {
                    centroids[(d, 0)] = (1.0 - (-1.0f64).powi(d as i32)) / 2.0;
                    centroids[(d, 1)] = 0.5;
                    centroids[(d, 2)] = (1.0 + (-1.0f64).powi(d as i32)) / 2.0;
                }
            }
            CentroidStrategy::KMeans => {
                // Not implemented
            }
        }

        let debug = false;

        if debug {
            println!("Centroids matrix:\n{}", centroids);
        }
    }
}
